universe/Condition.cpp

```diff
@@ -9956,6 +9965,252 @@ unsigned int Not::GetCheckSum() const {
     return retval;
 }
 
+///////////////////////////////////////////////////////////
+// OrderedAlternativesOf
+///////////////////////////////////////////////////////////
+OrderedAlternativesOf::OrderedAlternativesOf(
+    std::vector<std::unique_ptr<ConditionBase>>&& operands) :
+    ConditionBase(),
+    m_operands(std::move(operands))
+{}
+
+bool OrderedAlternativesOf::operator==(const ConditionBase& rhs) const {
+    if (this == &rhs)
+        return true;
+    if (typeid(*this) != typeid(rhs))
+        return false;
+
+    const OrderedAlternativesOf& rhs_ = static_cast<const OrderedAlternativesOf&>(rhs);
+
+    if (m_operands.size() != rhs_.m_operands.size())
+        return false;
+    for (unsigned int i = 0; i < m_operands.size(); ++i) {
+        CHECK_COND_VREF_MEMBER(m_operands.at(i))
+    }
+
+    return true;
+}
+
+void OrderedAlternativesOf::Eval(const ScriptingContext& parent_context,
+                                 ObjectSet& matches, ObjectSet& non_matches,
+                                 SearchDomain search_domain/* = NON_MATCHES*/) const
+{
+    std::shared_ptr<const UniverseObject> no_object;
+    ScriptingContext local_context(parent_context, no_object);
+
+    if (m_operands.empty()) {
+        ErrorLogger() << "OrderedAlternativesOf::Eval given no operands!";
+        return;
+    }
+    for (auto& operand : m_operands) {
+        if (!operand) {
+            ErrorLogger() << "OrderedAlternativesOf::Eval given null operand!";
+            return;
+        }
+    }
+
+    // Expressions like  'NOT OrderedAlternativesOf [ A B C ]'
+    // matches all candidates which do not match the topmost condition which has matches
+    // which means the something like
+    // OrderedAlternativesOf [
+    //     And [
+    //         Number low = 1 condition = And [
+    //             Object id = OuterCandidate.ID
+    //             A
+    //         ]
+    //         Not A
+    //     ]
+    //     And [
+    //         Number low = 1 condition = And [
+    //             Object id = OuterCandidate.ID
+    //             B
+    //         ]
+    //         Not B
+    //     ]
+    //     And [
+    //         Number low = 1 condition = And [
+    //             Object id = OuterCandidate.ID
+    //             C
+    //         ]
+    //         Not C
+    //     ]
+    // ]
+    if (search_domain == NON_MATCHES) {
+        // check all operand conditions on all objects in the non matches set
```

Comment:
Need to check candidates in matches and non_matches, no just in matches, to determine which subcondition to use. This should be the same regardless of the `search_domain`.

Reply:
Not sure what you refer to. But i already did in both cases. In the non_matches case i only have to restore the original state after figuring out the first matching operand condition

Reply:
probably i need to reread your comment and rethink

Reply:
Ok. So what would be a testcase?
```
And [
    Not Ship
    OrderedAlternativesOf [
        Ship
        Building
    ]
]
```
Hm. Lets say there are buildings and ships in the candidates. So you would say the Ship operand condition of OrderedAlternativesOf should be evalutated? So in such a case the All condition will always match nothing. 

Or should the OrderedAlternativesOf be context sensitive and fail the Ship operand condition? I think I would actually prefer this behavior.

Reply:
> ... the Ship operand condition of OrderedAlternativesOf should be chosen for evaluation?

Yes; the result of `And [A B]` should be the same as the result of `And [B A]`.

But more so, conditions like `And` can pass candidate sets as `matches` or `non_matches` and adjust `search_domain` as an implementation / optimization detail. The subcondition (in this case `OrderedAlternativesOf`) should match each candidate in the candidate set corresponding to the `search_domain` with the same criteria, regardless of which `search_domain` is specified.

For your example, you could rewrite it
````
OrderedAlternativesOf [
    And [Ship Not Ship]
    Building
]
````
Or just `Building`

Reply:
Ok, I will see that non_matches and matches get searched. I think I have  probably can skip the matching generation in the other case (e.g. when searching non_matches in the MATCHES domain).

Correct me, but this means that for deciding the alternative, all LocalCandidates are used, no matter what the other conditions do, right?

Concerning the example, of course using Ship and Not Ship does not make sense but it could make sense to have different conditions which sometimes leads to a similar situation. (So the just `Building` rewrite would miss the point.)

I guess a more exact rewrite for the generic case would be
```
OrderedAlternativesOf [
    And [ Ship [[MAYBE_NOT_SHIP]] ]
    And [ Building [[MAYBE_NOT_SHIP]] ]
]
```

Reply:
Meh, so this caused the combatTarget conditions to fail, they get pretty bloated now.

Reply:
> Meh, so this caused the combatTarget conditions to fail

Did you fix that?

Reply:
Yes, fixed. I am not too happy with the situation. 
```
    combatTargets = OrderedAlternativesOf [
        // Target Bombers and Heavy Bombers first
        And [
            [[COMBAT_TARGETS_VISIBLE_ENEMY]]
            Fighter
            Or [
               DesignHasPart name = "FT_HANGAR_3"
               DesignHasPart name = "FT_HANGAR_4"
            ]
        ]
        And [ [[COMBAT_TARGETS_VISIBLE_ENEMY]]  Fighter ]
        // if no fighters: target enemy ships
        And [ [[COMBAT_TARGETS_VISIBLE_ENEMY]]  [[COMBAT_TARGETS_NOT_DESTROYED_SHIP]] ]
    ]
```
I was entertaining the thought that future-me will implement some WithFilteredCandidates condition like the following:
```
    combatTargets = WithFilteredCandidates 
        scope = [[COMBAT_TARGETS_VISIBLE_ENEMY]]
        condition = OrderedAlternativesOf [
            // Target Bombers and Heavy Bombers first
            And [
                Fighter
                Or [
                   DesignHasPart name = "FT_HANGAR_3"
                   DesignHasPart name = "FT_HANGAR_4"
                ]
            ]
            Fighter
            // if no fighters: target enemy ships
            [[COMBAT_TARGETS_NOT_DESTROYED_SHIP]]
        ]
```
So for the purposes of this PR, it is fixed and tested.